### PR TITLE
SPIKE DRAFT: provenance and SBOM generation

### PR DESCRIPTION
**Pull request by stephenfuqua**
This is not for merging - only for review and discussion. For context, please see [Supply Chain Integrity](https://techdocs.ed-fi.org/display/EDFIODS/Supply+Chain+Integrity) in Tech Docs.

Most of the Actions yml files are renamed to disable them during my testing. I tested on my fork, where I could merge to main and create releases as often as I wanted.

End result: see this fake release, [v1.2.5](https://github.com/stephenfuqua/Ed-Fi-ODS-AdminApp/releases/tag/v1.2.5), which has SBOM manifest and provenance jsonl files attached to the release. In the [actions tab](https://github.com/stephenfuqua/Ed-Fi-ODS-AdminApp/actions), you can see that this latest run was generated from a tag rather than a branch. The tag was in turn created from a new GitHub release. So that's the workflow: create a release, and the provenance and SBOM files will be attached to the release. And [here's the action log](https://github.com/stephenfuqua/Ed-Fi-ODS-AdminApp/actions/runs/3053577102) for the release.

image

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -52,44 +79,137 @@ jobs:
...
-          }
-          $arguments.PackageFile = $artifact
-          ./build.ps1 Push @arguments
+          # sha256sum returns "<hashcode>  <name of file". Split that and return only the <hashcode>.
+          sbom_hash=$(sha256sum ./manifest/${{ env.MANIFEST_FILE }} | awk '{split($0,a); print a[1]}')
+          echo "::set-output name=sbom-hash::$sbom_hash"
+
+  sbom-release:
+    runs-on: ubuntu-latest
+    needs: [sbom-create]
```
Comment (andonyns) [inline .github/workflows/publish.yml:136]: This can go without the brackets as the rest of jobs

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -52,44 +79,137 @@ jobs:
...
+
+  provenance:
+    needs: pack
+    permissions:
+      actions: read
+      id-token: write
+      contents: write
+    uses: slsa-framework/slsa-github-generator/.github/workflows/generator_generic_slsa3.yml@v1.2.0
+    # This is failing :-(
+    # uses: ed-fi-alliance-sso/slsa-github-generator/.github/workflows/generator_generic_slsa3.yml@v1.2.0
```
Comment (andonyns) [inline .github/workflows/publish.yml:163]: The path here should be ed-fi-alliance-oss/slsa-github-generator/.github/workflows/generator_generic_slsa3.yml@v1.2.0

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -52,44 +79,137 @@ jobs:
...
     runs-on: ubuntu-latest
-
-    defaults:
-      run:
-        shell: pwsh
-
+    needs: pack
+    permissions:
+      actions: read
+      id-token: write
```
Comment (andonyns) [inline .github/workflows/publish.yml:87]: Is this permission required? I tested this on my branch without the permission and the workflow finished correctly. Wondering for this one and the pack and provenance jobs

File: .github/workflows/publish.yml [hunk 4]
```diff
@@ -17,32 +19,57 @@ env:
...
+  MANIFEST_FILE: "_manifest/spdx_2.2/manifest.spdx.json"
 
 jobs:
   pack:
-
     runs-on: ubuntu-latest
-
     defaults:
       run:
         shell: pwsh
```
Comment (andonyns) [inline .github/workflows/publish.yml:30]: Now that there are a few scripts on bash, I wonder if it is better to default to bash (the original) and specify PowerShell on the required scripts.

File: .github/workflows/publish.yml (hunk 2 above)
Comment (andonyns) [reply .github/workflows/publish.yml:163]: I tested this on my fork and it worked correctly

Comment (stephenfuqua) [reply .github/workflows/publish.yml:163]: Oh, I see that I wrote "sso" instead of "oss" 😆

File: .github/workflows/publish.yml (hunk 3 above)
Comment (stephenfuqua) [reply .github/workflows/publish.yml:87]: I didn't test this carefully after copying some permissions from the SLSA action. Thinking about it, this definitely should _not_ be needed.

File: .github/workflows/publish.yml [hunk 5]
```diff
@@ -17,32 +19,57 @@ env:
...
+    outputs:
+      hashes: ${{ steps.hashes.outputs.hashes }}
 
     steps:
     - uses: actions/checkout@2541b1294d2704b0964813337f33b291d3f8596b # v3.0.2
+
```
[the remark this answers is not in the document]
Comment (andonyns) [reply .github/workflows/publish.yml:36]: I really like that, for the future, it might be good to add a linting verification/auto-fix for yml files :)

**Pull request closed without merging**